### DISCUSSION TITLE
Chunk Utils not working?

### DISCUSSION THREAD

**Discussion by GoldenGamerLP** [Minestom/Minestom#1444] [Help]
Hello, quick question, i try to load chunks in a specific range, but the chunks always are DynamicChunk(0:0)

```
long[] chunks = new long[1024];
        AtomicInteger counter = new AtomicInteger();
        ChunkUtils.forChunksInRange(position, Math.max(this.lowerCorner().chunkX() - this.upperCorner().chunkX(), this.lowerCorner().chunkZ() - this.upperCorner().chunkZ()),
                (v1, v2) -> {
                    chunks[counter.getAndIncrement()] = ChunkUtils.getChunkIndex(v1,v2);
                    Logger.getAnonymousLogger().info(ChunkUtils.getChunkIndex(v1,v2) + "");
                });

        ChunkUtils.optionalLoadAll(instance, chunks, chunk -> Logger.getAnonymousLogger().info(chunk.toString())).thenRunAsync(() -> Logger.getAnonymousLogger().info("Chunks 0 > " + (System.currentTimeMillis() - ms)));
```

**Comment by KrystilizeNevaDies** [comment]
It would help to simplify this, break the issue down.
Firstly just ensure that the chunk indexes are correct.
Then do a single chunk retrieval and ensure that it's the correct chunk.

Doing multiple things at once here makes it much harder to discern what could/would go wrong.

**Reply by GoldenGamerLP** [reply]
Alright, ive found my mistake, in my code i used Math.max(...), which returned a negative number, i used pow and sqrt to get a positive number, thats my first issue fixed.

Secondly, i think a found another issue.
![image](https://user-images.githubusercontent.com/65727384/195884942-9d75e720-2492-474a-971c-2f11fbd58c81.png)

Here you can see it goes from the range as a negative number all the way to the positive range number. But the consumer only adds them. (-+ = +) Or is this an other error than i think, because i get for the range "20" exactly 1681 chunks.

**Reply by GoldenGamerLP** [reply]
Nevermind, the chunk loading doesnt work anymore, it was not the negative number

**Reply by GoldenGamerLP** [reply]
Push

I tried using the chunkbatch map, converting the long into chunk coordinates and loading them, the chunks did not load either. The result stayed the same. All chunks did not load, wether i tried to convert the long values into chunks and load them or iterate through the coordinates myself and load them. Ive always got the error "cannot apply chunk batch, chunk is not loaded"

**Reply by mworzala** [reply]
As small as possible of a reproducible example would be best. If that cannot be done for some reason then the exact code you're using.

**Reply by GoldenGamerLP** [reply]
Here are some snippets:

Main programm:

```
        //debug
        long ms = System.currentTimeMillis();
        Logger.getAnonymousLogger().info("Normal 0 " + (System.currentTimeMillis() - ms));


        List<CompletableFuture<?>> blocks = new ArrayList<>();
        int mid = regionBlocks.size() / 2;

        //first list
        blocks.add(CompletableFuture.runAsync(() -> {
            List<Region.Block> first = regionBlocks.subList(0, mid);
            goThroughBlocks(position, blockBatch, first);
        }));

        //second half of list
        blocks.add(CompletableFuture.runAsync(() -> {
            List<Region.Block> second = regionBlocks.subList(mid + 1, regionBlocks.size());
            goThroughBlocks(position, blockBatch, second);
        }));

        CompletableFuture<?> chunks = ChunkUtils
                .optionalLoadAll(
                        instance,
                        blockBatch.loadableChunks(),
                        null
                );

        //Adding all non depending tasks
        //Only after all blocks iteration and chunk loading!
        CompletableFuture.allOf(blocks.toArray(CompletableFuture[]::new)).thenRunAsync(chunks::join)
                .thenRun(() -> blockBatch.apply(instance, () -> future.complete(new Region(instance, position, position.add(width, height, length))))).whenComplete((unused, throwable) -> {
                    if(throwable != null) throwable.printStackTrace();
                });
        Logger.getAnonymousLogger().info("Normal 1 " + (System.currentTimeMillis() - ms));
        return future;
```

My modified AbsolutBlockBatch method:

```
    @Experimental()
    public long[] loadableChunks() {
        return this.chunkBatchesMap.keySet().toArray(new long[0]);
    }
```

**Reply by GoldenGamerLP** [reply]
> As small as possible of a reproducible example would be best. If that cannot be done for some reason then the exact code you're using.

I hope thats enough

**Reply by mworzala** [reply]
I will try to look into it after work tonight

**Reply by GoldenGamerLP** [reply]
#Push

**Reply by GoldenGamerLP** [reply]
@mworzala
